module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,207 @@
+package indexer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/migrations"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// ExecutionState indexes the execution state.
+type ExecutionState struct {
+	registers  storage.RegisterIndex
+	headers    storage.Headers
+	events     storage.Events
+	indexRange *SequentialIndexRange
+	log        zerolog.Logger
+}
+
+// New execution state indexer used to ingest block execution data and index it by height.
+// The passed RegisterIndex storage must be populated to include the first and last height otherwise the indexer
+// won't be initialized to ensure we have bootstrapped the storage first.
+func New(
+	log zerolog.Logger,
+	registers storage.RegisterIndex,
+	headers storage.Headers,
+	events storage.Events,
+) (*ExecutionState, error) {
+	// we must have the first height set in the storage otherwise error out
+	first, err := registers.FirstHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing first height")
+	}
+
+	// we must have the latest height set in the storage otherwise error out
+	last, err := registers.LatestHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing latest height")
+	}
```

Comment:
this logic happens in `NewRegisters()` is called, so we don't need this check here. we're guranteed these values when the registers object is able to initialize

Reply:
I see it the other way around actually, the fact the check is done in storage is an extra step. This to me feels like it should live inside the indexer as it's right in the domain logic of the indexer. I would argue the storage wouldn't need this check but I guess it was chosen to have it to be reusable by other consumers.

Reply:
https://github.com/onflow/flow-go/pull/4653#discussion_r1335681180